File: algorithm_datastructure/leetcode/69.sqrt-x.py
#
# @lc app=leetcode id=69 lang=python3
#
# [69] Sqrt(x)
#

# @lc code=start
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0
        left = 0
        right = x
        sqrt = x
        # Uses left close and right close
        while left <= right:
            mid = (left + right) // 2
            if mid ** 2 == x:
                return mid
            elif mid ** 2 < x:
                left = mid + 1
                sqrt = mid
            else:
                right = mid - 1

        return sqrt


# @lc code=end

File: algorithm_data_structure/leetcode/69.sqrt-x.cpp
/*
 * @lc app=leetcode id=69 lang=cpp
 *
 * [69] Sqrt(x)
 */

// @lc code=start

#include <stdint.h>

class Solution
{
public:
    int mySqrt(int x)
    {
        if (x < 0)
        {
            return x;
        }

        // Use left close and right close
        int left = 1, right = (x / 2) + 1, mid = 0;
        uint64_t mid_q = 0;

        while (left <= right)
        {
            mid = left + (right - left) / 2; // Avoid overflow

            mid_q = (uint64_t)mid * (uint64_t)mid;

            if (mid_q == x)
            {
                return mid;
            }
            else if (mid_q < x)
            {
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
        }

        return right;
    }
};
// @lc code=end
